In [1]:
import numpy as np
import pandas as pd

import torch
torch.set_printoptions(edgeitems=5, precision=2, linewidth=130)   # устанавливаем параметры вывода тензоров на печать под себя

import csv

In [2]:
wine_path = "data/winequality-white.csv"

# грузим табличные данные

## pandas

In [3]:
ddf = pd.read_csv(wine_path, sep=";", dtype=np.float32)
ddf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float32
 1   volatile acidity      4898 non-null   float32
 2   citric acid           4898 non-null   float32
 3   residual sugar        4898 non-null   float32
 4   chlorides             4898 non-null   float32
 5   free sulfur dioxide   4898 non-null   float32
 6   total sulfur dioxide  4898 non-null   float32
 7   density               4898 non-null   float32
 8   pH                    4898 non-null   float32
 9   sulphates             4898 non-null   float32
 10  alcohol               4898 non-null   float32
 11  quality               4898 non-null   float32
dtypes: float32(12)
memory usage: 229.7 KB


In [4]:
ddf.map(lambda d: round(d,4))

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0   1.0010  3.00       0.45   
1                    14.0                 132.0   0.9940  3.30       0.49   
2                    30.0                  97.0   0.9951  3.26       0.44   
3                    47.0                 186.0   0.9956  3.19       0.40   
4                    47.0                 186.0   0.9956  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0   0.9911  3.27       0.50   
4894                 57.0                 168.0   0.9949  3.15       0.46   
4895                 30.0                 111.0   0.9925  2.99       0.46   
4896                 20.0                 110.0   0.9887  3.34       0.38   
4897                 22.0                  98.0   0.9894  3.26       0.32   

      alcohol  quality  
0         8.8      6.0  
1         9.5      6.0  
2        10.1      6.0  
3         9.9      6.0  
4         9.9      6.0  
...       ...      ...  
4893     11.2      6.0  
4894      9.6      5.0  
4895      9.4      6.0  
4896     12.8      7.0  
4897     11.8      6.0  

[4898 rows x 12 columns]

## numpy

In [5]:
wineq_numpy = np.loadtxt(wine_path, dtype=np.float32, delimiter=";", skiprows=1)
wineq_numpy

array([[ 7.  ,  0.27,  0.36, ...,  0.45,  8.8 ,  6.  ],
       [ 6.3 ,  0.3 ,  0.34, ...,  0.49,  9.5 ,  6.  ],
       [ 8.1 ,  0.28,  0.4 , ...,  0.44, 10.1 ,  6.  ],
       ...,
       [ 6.5 ,  0.24,  0.19, ...,  0.46,  9.4 ,  6.  ],
       [ 5.5 ,  0.29,  0.3 , ...,  0.38, 12.8 ,  7.  ],
       [ 6.  ,  0.21,  0.38, ...,  0.32, 11.8 ,  6.  ]],
      shape=(4898, 12), dtype=float32)

## python csv

In [6]:
csv_lineread_generator = csv.reader(open(wine_path), delimiter=';')  # создаем генератор для чтения csv-файла по строкам

In [7]:
manual_ddf = pd.DataFrame(
    columns=next(csv_lineread_generator),                                      # читаем следующую (исторически первую) строку
    data   =np.array([np.float32(row) for row in csv_lineread_generator])      # читаем остальные строки до конца файла
)
manual_ddf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float32
 1   volatile acidity      4898 non-null   float32
 2   citric acid           4898 non-null   float32
 3   residual sugar        4898 non-null   float32
 4   chlorides             4898 non-null   float32
 5   free sulfur dioxide   4898 non-null   float32
 6   total sulfur dioxide  4898 non-null   float32
 7   density               4898 non-null   float32
 8   pH                    4898 non-null   float32
 9   sulphates             4898 non-null   float32
 10  alcohol               4898 non-null   float32
 11  quality               4898 non-null   float32
dtypes: float32(12)
memory usage: 229.7 KB


In [8]:
print('\nFYI: кол-во несовпадений между двумя способами вытащить таблицу из файла =', (ddf != manual_ddf).sum().sum() )


FYI: кол-во несовпадений между двумя способами вытащить таблицу из файла = 0


## torch

In [9]:
wineq = torch.from_numpy(wineq_numpy)

wineq.shape, wineq.dtype

(torch.Size([4898, 12]), torch.float32)

# типы данных в ML

* НЕПРЕРЫВНЫЕ
  * есть относительная метрика (не всегда)
  * есть интервальная метрика (всегда)
  * есть метрика порядка (разные)
* УПОРЯДОЧЕННЫЕ
  * есть метрика порядка
* КАТЕГОРИАЛЬНЫЕ
  * есть только номинальная шкала

# собственно данные и целевые переменные на них

In [10]:
data = wineq[:, :-1]            # выбираем все строки и все столбцы, кроме последнего
data, data.shape

(tensor([[7.00e+00, 2.70e-01, 3.60e-01, 2.07e+01, 4.50e-02,  ..., 1.70e+02, 1.00e+00, 3.00e+00, 4.50e-01, 8.80e+00],
         [6.30e+00, 3.00e-01, 3.40e-01, 1.60e+00, 4.90e-02,  ..., 1.32e+02, 9.94e-01, 3.30e+00, 4.90e-01, 9.50e+00],
         [8.10e+00, 2.80e-01, 4.00e-01, 6.90e+00, 5.00e-02,  ..., 9.70e+01, 9.95e-01, 3.26e+00, 4.40e-01, 1.01e+01],
         [7.20e+00, 2.30e-01, 3.20e-01, 8.50e+00, 5.80e-02,  ..., 1.86e+02, 9.96e-01, 3.19e+00, 4.00e-01, 9.90e+00],
         [7.20e+00, 2.30e-01, 3.20e-01, 8.50e+00, 5.80e-02,  ..., 1.86e+02, 9.96e-01, 3.19e+00, 4.00e-01, 9.90e+00],
         ...,
         [6.20e+00, 2.10e-01, 2.90e-01, 1.60e+00, 3.90e-02,  ..., 9.20e+01, 9.91e-01, 3.27e+00, 5.00e-01, 1.12e+01],
         [6.60e+00, 3.20e-01, 3.60e-01, 8.00e+00, 4.70e-02,  ..., 1.68e+02, 9.95e-01, 3.15e+00, 4.60e-01, 9.60e+00],
         [6.50e+00, 2.40e-01, 1.90e-01, 1.20e+00, 4.10e-02,  ..., 1.11e+02, 9.93e-01, 2.99e+00, 4.60e-01, 9.40e+00],
         [5.50e+00, 2.90e-01, 3.00e-01, 1.10e+00, 

In [11]:
target = wineq[:, -1]           # выбираем все строки и последний столбец
target, target.shape

(tensor([6., 6., 6., 6., 6.,  ..., 6., 5., 6., 7., 6.]), torch.Size([4898]))

## кодирование разметки

### номера классов

In [12]:
target = wineq[:, -1].long()
target

tensor([6, 6, 6, 6, 6,  ..., 6, 5, 6, 7, 6])

In [13]:
print('\nFYI: кол-во неправильных округлений при конверсии типов =', ((target - wineq[:, -1]) * (target - wineq[:, -1])).sum().numpy())


FYI: кол-во неправильных округлений при конверсии типов = 0.0


**Хранение оценок вкусовых качеств вина в целочисленном векторе влечет их упорядоченность, что в данном случае может быть неуместно, ведь оценка 1 ниже оценки 4. Это также подразумевает наличие расстояния между оценками: то есть расстояние между 1 и 3 равно расстоянию между 2 и 4. Если для нашей величины это подходит — замечательно.**

**Если же, напротив, оценки совершенно дискретны, например, означают сорт винограда, то ohe-кодирование подойдет намного лучше,
поскольку не подразумевает упорядоченности или расстояния**

### *ohe* используя метод `tensor.scatter_`

Syntax: `self_tensor.scatter_(axis, index, src)`
* `axis` (int) – the axis along which to index
* `index` (LongTensor) – the indices of elements to scatter, should have same number of dimensionalities as `self_tensor`
* `src` (Tensor) – the source of element(s) to scatter

**параметры избыточны, но это дает дополнительный чек**

In [14]:
target_onehot = torch.zeros(target.shape[0], 10)
target_onehot

tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        ...,
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [15]:
target_unsqueezed = target.long().unsqueeze(1)
target_unsqueezed

tensor([[6],
        [6],
        [6],
        [6],
        [6],
        ...,
        [6],
        [5],
        [6],
        [7],
        [6]])

In [16]:
target_onehot.scatter_(1, target_unsqueezed, 1.0)

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        ...,
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.]])

**Пояснение:** аргументы метода scatter_ следующие:
* измерение, по которому указываются следующие два аргумента;
* тензор-столбец с индексами заполняемых элементов;
* тензор с элементами для распределения или одно заполняющее скалярное
значение (в данном случае 1).

Другими словами, предыдущий вызов гласит: «Для каждой строки нужно воспользоваться индексом целевой метки (совпадающим с оценкой в нашем случае) в качестве индекса столбца для задания значения 1,0». В результате получается тензор, кодирующий категориальную информацию.

Размерность второго аргумента метода scatter_, тензора индексов, должна совпадать с заполняемым тензором. Поскольку размерность тензора target_onehot равна 2 (4898 × 10), необходимо добавить еще одно фиктивное измерение в target с помощью функции unsqueeze

**в отличие от `inline=True` пандас, в торче self-методы и self меняют (ниже) и его же возвращают (выше)**

In [17]:
target_onehot

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        ...,
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.]])

# нормализация данных

## вектор средних

In [18]:
data_mean = torch.mean(data, dim=0)
data_mean

tensor([6.85e+00, 2.78e-01, 3.34e-01, 6.39e+00, 4.58e-02, 3.53e+01, 1.38e+02, 9.94e-01, 3.19e+00, 4.90e-01, 1.05e+01])

## вектор дисперсий

In [19]:
data_var = torch.var(data, dim=0)
data_var

tensor([7.12e-01, 1.02e-02, 1.46e-02, 2.57e+01, 4.77e-04, 2.89e+02, 1.81e+03, 8.95e-06, 2.28e-02, 1.30e-02, 1.51e+00])

## нормируем

In [20]:
data_normalized = (data - data_mean) / torch.sqrt(data_var)

## проверяем

In [21]:
torch.mean(data_normalized, dim=0)

tensor([-1.81e-07,  2.16e-07,  6.31e-08,  4.13e-08,  5.33e-08, -8.72e-08,  2.49e-08, -5.45e-08,  8.35e-07,  4.35e-07,  4.50e-08])

**вот она, точность `float32`!**

In [22]:
torch.var(data_normalized, dim=0)

tensor([1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00])

**кажется, что VAR и STD всегда получаются точно 1, а вот средние не так точны, но это трюки округления, которые видны, если вычесть 1.0:**

In [23]:
torch.var(data_normalized, dim=0) - 1.0

tensor([-5.96e-08,  0.00e+00, -1.19e-07,  0.00e+00,  0.00e+00,  0.00e+00,  1.19e-07,  0.00e+00, -5.96e-08,  0.00e+00,  1.19e-07])

**а также порядка вычислений, при которых постоянно идут округления и накапливается ошибка:**

In [24]:
data_normalized = data / torch.sqrt(data_var) - data_mean / torch.sqrt(data_var)  # немного другой порядок действий

torch.var(data_normalized, dim=0) - 1.0

tensor([-5.96e-08,  0.00e+00, -1.19e-07,  0.00e+00,  1.19e-07,  0.00e+00,  1.19e-07,  5.96e-07, -5.96e-08,  0.00e+00,  3.58e-07])

# NB!
PyTorch дает возможность непосредственно использовать индексы классов в качестве целей при обучении нейронной сети. Однако для использования оценок в качестве категориального входного сигнала сети необходимо преобразовать его в ohe-кодированный тензор.

# Еще без ML: поиск пороговых значений руками

In [25]:
bad_indexes = (target <= 3)                                 # порог - это бинарность
bad_indexes.shape, bad_indexes.dtype

(torch.Size([4898]), torch.bool)

In [26]:
print('\nFYI: кол-во плохих вин =', bad_indexes.sum().numpy(), '\n')


FYI: кол-во плохих вин = 20 



In [27]:
bad_data = data[bad_indexes]
bad_data

tensor([[8.50e+00, 2.60e-01, 2.10e-01, 1.62e+01, 7.40e-02, 4.10e+01, 1.97e+02, 9.98e-01, 3.02e+00, 5.00e-01, 9.80e+00],
        [5.80e+00, 2.40e-01, 4.40e-01, 3.50e+00, 2.90e-02, 5.00e+00, 1.09e+02, 9.91e-01, 3.53e+00, 4.30e-01, 1.17e+01],
        [9.10e+00, 5.90e-01, 3.80e-01, 1.60e+00, 6.60e-02, 3.40e+01, 1.82e+02, 9.97e-01, 3.23e+00, 3.80e-01, 8.50e+00],
        [7.10e+00, 3.20e-01, 3.20e-01, 1.10e+01, 3.80e-02, 1.60e+01, 6.60e+01, 9.94e-01, 3.24e+00, 4.00e-01, 1.15e+01],
        [6.90e+00, 3.90e-01, 4.00e-01, 4.60e+00, 2.20e-02, 5.00e+00, 1.90e+01, 9.92e-01, 3.31e+00, 3.70e-01, 1.26e+01],
        [1.03e+01, 1.70e-01, 4.70e-01, 1.40e+00, 3.70e-02, 5.00e+00, 3.30e+01, 9.94e-01, 2.89e+00, 2.80e-01, 9.60e+00],
        [7.90e+00, 6.40e-01, 4.60e-01, 1.06e+01, 2.44e-01, 3.30e+01, 2.27e+02, 9.98e-01, 2.87e+00, 7.40e-01, 9.10e+00],
        [8.30e+00, 3.30e-01, 4.20e-01, 1.15e+00, 3.30e-02, 1.80e+01, 9.60e+01, 9.91e-01, 3.20e+00, 3.20e-01, 1.24e+01],
        [8.60e+00, 5.50e-01, 3.50e-01, 1

## сравним статистики хороших, средних и плохих вин - по среднему значению, например

In [28]:
bad_data = data[target <= 3]
mid_data = data[(target > 3) & (target < 7)]
good_data = data[target >= 7]

bad_mean = torch.mean(bad_data, dim=0)
mid_mean = torch.mean(mid_data, dim=0)
good_mean = torch.mean(good_data, dim=0)

for i, args in enumerate(zip(ddf.columns, bad_mean, mid_mean, good_mean)):
    print('{:2} {:20} {:6.2f} {:6.2f} {:6.2f}'.format(i, *args))

 0 fixed acidity          7.60   6.89   6.73
 1 volatile acidity       0.33   0.28   0.27
 2 citric acid            0.34   0.34   0.33
 3 residual sugar         6.39   6.71   5.26
 4 chlorides              0.05   0.05   0.04
 5 free sulfur dioxide   53.33  35.42  34.55
 6 total sulfur dioxide 170.60 141.83 125.25
 7 density                0.99   0.99   0.99
 8 pH                     3.19   3.18   3.22
 9 sulphates              0.47   0.49   0.50
10 alcohol               10.34  10.26  11.42


**монотонность выявляет явную зависимость**

## сделаем пороговый классификатор по одной переменной - для прикола

In [29]:
total_sulfur_threshold = 141.83
total_sulfur_data = data[:,6]
predicted_indexes = torch.lt(total_sulfur_data, total_sulfur_threshold)

predicted_indexes.shape, predicted_indexes.dtype, predicted_indexes.sum()

(torch.Size([4898]), torch.bool, tensor(2727))

In [30]:
actual_indexes = target > 5

actual_indexes.shape, actual_indexes.dtype, actual_indexes.sum()

(torch.Size([4898]), torch.bool, tensor(3258))

In [31]:
n_matches = torch.sum(actual_indexes & predicted_indexes).item()
n_predicted = torch.sum(predicted_indexes).item()
n_actual = torch.sum(actual_indexes).item()

n_matches/len(ddf), n_matches / n_predicted, n_matches / n_actual

(0.412004899959167, 0.74000733406674, 0.6193984039287906)

**это знаменитая тройка `(Accuracy, Precision, Recall)` для нашего порогового классификатора, выявляющего плохие вина!**